### PR TITLE
Backend/feature/communities tests

### PR DESCRIPTION
**Pull request by aapeliv**
This is a mess

### CODE REVIEW

**Comment by aapeliv** [comment]
@kalvdans: this takes forever to run. Do you know how to only run the `testing_communities` fixture once in the file and run all tests against it, without recreating everything every time?

**Comment by kalvdans** [comment]
> @kalvdans: this takes forever to run. Do you know how to only run the `testing_communities` fixture once in the file and run all tests against it, without recreating everything every time?

Can we perhaps do a database dump the first time, and just reload the dump?

otherwise we just have to optimize, don't create stuff that's not needed by the active test.

**Comment by aapeliv** [comment]
It seems that `scope="module"` should do precisely this. However, I'm running into some issues with PostGIS now, I think this is doing something unexpected.

**Comment by kalvdans** [comment]
> It seems that `scope="module"` should do precisely this.

I have had a hard time understanding the magic what scope="module" does. If you want tests to be dependent on the leftover from an earlier test, use the `unittest` framework and its setUpClass function. https://docs.python.org/3/library/unittest.html#unittest.TestCase.setUpClass

pytest will run tests written in the unittest framework as well.

**Comment by aapeliv** [comment]
It just reuses the same fixture for all tests in the same module, without destroying it and recreating it, right?

I don't think we should mix in unittest, this seems to work fine?

I mean, it's not as nice as having completely isolated unit tests, but at the same time, these are read-only queries.

Another option would be to just put them all in the same test?

**Comment by aapeliv** [comment]
@kalvdans: what would you want me to do about this? I guess the lame option is to just have these independent tests that are very slow. For that, I can revert the last commit. Another is to share the same `testing_communities` fixture as it is now. Up to you?

**Comment by kalvdans** [comment]
> I guess the lame option is to just have these independent tests that are very slow. For that, I can revert the last commit. Another is to share the same `testing_communities` fixture as it is now. Up to you?

The issue with `scope="module"` is that if you have a test with the `db` fixture in between two tests with `testing_communities` fixture, it will destroy the database. It is better to use `scope="class"` and put all the tests that need to run after one another in that class. I'll try to see if I get something to work on top of your branch.

See https://github.com/Couchers-org/couchers/pull/554 for my take on the scoping.

**Comment by aapeliv** [comment]
@kalvdans: that PR looks very good to me. Thanks for doing that. Are you happy for me to merge that into this and this into `develop`?

File: app/backend/src/tests/test_groups.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        assert res.parents[3].group.description == "Description for Country 2, Region 1, Foodies"
+        assert res.main_page.type == pages_pb2.PAGE_TYPE_MAIN_PAGE
+        assert res.main_page.slug == "main-page-for-the-country-2-region-1-foodies-community"
+        assert res.main_page.last_editor_user_id == 6
+        assert res.main_page.creator_user_id == 6
+        assert res.main_page.owner_user_id == 6
+        assert res.main_page.title == "Main page for the Country 2, Region 1, Foodies community"
+        assert res.main_page.content == "There is nothing here yet..."
+        # assert res.main_page.can_edit # TODO
+        assert lists_equal(res.main_page.editor_user_ids, [6])
```
Comment (kalvdans) [inline app/backend/src/tests/test_groups.py]: ` assert res.main_page.editor_user_ids == [6]`

File: app/backend/src/tests/test_groups.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        hitchhikers_id = get_group_id(session, "Hitchhikers")
+        c1r2foodies_id = get_group_id(session, "Country 1, Region 2, Foodies")
+
+    with groups_session(token1) as api:
+        res = api.ListMembers(
+            groups_pb2.ListMembersReq(
+                group_id=hitchhikers_id,
+            )
+        )
+        assert lists_equal(res.member_user_ids, [user1_id, user2_id, user5_id, user8_id])
```
Comment (kalvdans) [inline app/backend/src/tests/test_groups.py]: `assert set(res.member_user_ids) == set([user1_id, user2_id, user5_id, user8_id])`

File: app/backend/src/tests/test_fixtures.py [hunk 3]
```diff
@@ -406,3 +431,8 @@ def fast_verify(hashed: bytes, password: bytes) -> bool:
     with patch("couchers.crypto.nacl.pwhash.verify", fast_verify):
         with patch("couchers.crypto.nacl.pwhash.str", fast_hash):
             yield
+
+
+def lists_equal(list1, list2):
+    """Compare equality of (hashable) elements in two lists, by computing size of symmetric set difference"""
+    return len(set(list1) ^ set(list2)) == 0
```
Comment (kalvdans) [inline app/backend/src/tests/test_fixtures.py]: Remove function and inline into callers instead - I've commented on some examples.

**Review by kalvdans — CHANGES_REQUESTED** [review]

File: app/backend/src/tests/test_communities.py [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+
+            res = api.ListNearbyUsers(
+                communities_pb2.ListNearbyUsersReq(
+                    community_id=c1r1c2_id,
+                )
+            )
+            assert lists_equal(res.nearby_user_ids, [user4_id])
+
+
+# TODO: requires transferring of content
```
Comment (kalvdans) [inline app/backend/src/tests/test_communities.py:478]: I'm usually against leaving TODO:s as comments in the code, I rather would like an issue opened. But I don't feel strongly about it so I'm ok with leaving it as-is.

Comment (aapeliv) [reply app/backend/src/tests/test_communities.py:478]: Yes, I was going to work on those next! (Implementation then tests)

**Review by kalvdans — APPROVED** [review]
I get some spurios
```
psycopg2.errors.InternalError_: no spatial operator found for 'st_contains': opfamily 21450 type 21337
```
test failures when running the whole test suite locally. But if it doesn't occur on CI, I'm fine with how this branch looks now!

**Pull request merged**